File: bench/triad.py
"""The unoptimised triad: what each of the three formats looks like before anyone tunes it.

This is level zero of the ladder, and it is a deliberate choice rather than a limitation. Every
optimisation the project goes on to measure is a delta against these three, so they have to be the
honest defaults — what you get without thinking about it:

- **baseline** — an HTTP verb and a base URL. No documentation of any kind.
- **mcp** — every schema declared on connect, responses returned whole. This is the off-the-shelf
  server, not a caricature of one.
- **cli** — one command tool, discovery through `--help`, responses returned whole. Nobody pastes a
  manual into a system prompt unprompted.

The variants already implemented elsewhere in this package — deferred loading, a preloaded manual,
result filtering — are not part of this level. They are the optimisation catalogue, and their value
is only legible as a distance from here.
"""

from __future__ import annotations

from dataclasses import dataclass

from bench.arms import Disclosure, Handling


@dataclass(frozen=True, slots=True)
class Arm:
    """One configuration to run, with the turn ceiling it needs to be judged fairly."""

    key: str
    fmt: str
    disclosure: Disclosure
    handling: Handling
    max_turns: int
    why: str


TRIAD: tuple[Arm, ...] = (
    Arm(
        key="baseline",
        fmt="raw",
        disclosure="lazy",
        handling="whole",
        # Blind discovery needs room. Cut it at the same ceiling as the documented arms and the
        # result would measure the ceiling instead of the arm; running out of turns here is a
        # finding, and it is recorded as one.
        max_turns=30,
        why="an HTTP verb and a base URL: the price tag on having no documentation at all",
    ),
    Arm(
        key="mcp",
        fmt="mcp",
        disclosure="eager",
        handling="whole",
        max_turns=12,
        why="the off-the-shelf server: every schema up front, responses whole",
    ),
    Arm(
        key="cli",
        fmt="cli",
        disclosure="lazy",
        handling="whole",
        max_turns=12,
        why="one command tool, discovery through --help, responses whole",
    ),
)


def by_key(key: str) -> Arm:
    for arm in TRIAD:
        if arm.key == key:
            return arm
    raise KeyError(f"unknown arm: {key!r}")

File: src/triad.ts
/**
 * The unoptimised triad: what each of the three formats looks like before anyone tunes it.
 *
 * This is level zero of the ladder, and a deliberate choice rather than a limitation. Every
 * optimisation the project goes on to measure is a delta against these three, so they have to be the
 * honest defaults — what you get without thinking about it.
 *
 * Every arm gets unrestricted `bash`. That is the common denominator a real terminal agent already
 * has; the arm is what gets installed beside it. In dsh the tool surface is exactly what the harness
 * mounts — bash for every arm, plus one MCP client per installed server — so an arm needs no tool
 * allowlist of its own: installing the server is what puts `mcp__<server>__*` in front of the model.
 */

import type { Arm } from "./harness.ts";

const REPO = new URL("..", import.meta.url).pathname.replace(/^\/([A-Za-z]:)/, "$1");

export const TRIAD: readonly Arm[] = [
  {
    key: "baseline",
    mcpServers: {},
    installs: [],
    pathAdditions: [],
    why: "a shell and nothing else: the price of having no documentation at all",
  },
  {
    // The server is spawned by the client on the host, as in any real installation — which places
    // it outside the sandbox; declared in .claude/reference/arm-symmetry.md.
    key: "mcp",
    mcpServers: { camara: `bun run ${REPO}src/artifacts/mcp-server.ts` },
    installs: [],
    pathAdditions: [],
    why: "the user installed an MCP server; the client decides how it reaches the model",
  },
  {
    // The binary is copied into the trial's cwd: it is the installation, and it has to exist
    // inside the sandbox, where the repository does not.
    key: "cli",
    mcpServers: {},
    installs: [{ from: `${REPO}bin/api`, to: "bin/api" }],
    pathAdditions: ["bin"],
    why: "the user installed a CLI; the shell it already had is how it gets called",
  },
];

export function armByKey(key: string): Arm {
  const found = TRIAD.find((arm) => arm.key === key);
  if (!found) throw new Error(`unknown arm: '${key}'`);
  return found;
}
